flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java

```diff
@@ -409,39 +405,31 @@ private void discardLocalStateForCheckpoint(long checkpointID, Optional<TaskStat
                     }
                 });
 
-        Optional<LocalRecoveryDirectoryProvider> directoryProviderOptional =
-                localRecoveryConfig.getLocalStateDirectoryProvider();
+        File checkpointDir = getCheckpointDirectory(checkpointID);
 
-        if (directoryProviderOptional.isPresent()) {
-            File checkpointDir =
-                    directoryProviderOptional
-                            .get()
-                            .subtaskSpecificCheckpointDirectory(checkpointID);
```

Comment:
This slightly changes the semantics - it now tries to re-create the folder if it doesn't exist. Is it intentional? If not, is it safe?

(asserting that the provider exists looks safe)

Reply:
> Is it intentional? If not, is it safe?

Yes, this slightly changes the semantics, it is to reuse the logic of `getCheckpointDirectory`. It is safe, because the directory would be deleted by:
```
 try {
            deleteDirectory(checkpointDir);
        } catch (IOException ex) {
...
}
```

I will refactor the ` getCheckpointDirectory()` in https://issues.apache.org/jira/browse/FLINK-28614 , separate the logic of `get` and `create`.